File: sdks/typescript/src/jwt_values.ts
/** A2A JWT configuration and claims values without owning token verification. */
export class JwtConfig {
  private constructor(
    private readonly algorithm: "HS256" | "RS256" | null,
    private readonly verificationKey: string | null,
    public readonly issuer: string | null,
    public readonly audience: string | null,
  ) {}

  public static hs256(secret: string): JwtConfig {
    if (typeof secret !== "string") throw new TypeError("JWT secret must be text");
    return new JwtConfig("HS256", secret, null, null);
  }

  public static rs256(publicKey: string): JwtConfig {
    if (typeof publicKey !== "string" || !looksLikeRsaPublicKey(publicKey)) {
      throw new Error("invalid RSA public key");
    }
    return new JwtConfig("RS256", publicKey, null, null);
  }

  public static disabled(): JwtConfig {
    return new JwtConfig(null, null, null, null);
  }

  public withIssuer(issuer: string): JwtConfig {
    if (typeof issuer !== "string") throw new TypeError("JWT issuer must be text");
    return new JwtConfig(this.algorithm, this.verificationKey, issuer, this.audience);
  }

  public withAudience(audience: string): JwtConfig {
    if (typeof audience !== "string") throw new TypeError("JWT audience must be text");
    return new JwtConfig(this.algorithm, this.verificationKey, this.issuer, audience);
  }

  public isEnabled(): boolean {
    return this.algorithm !== null;
  }

  public toString(): string {
    return `JwtConfig { enabled: ${this.isEnabled()}, algorithm: ${this.algorithm ?? "none"}, issuer: ${this.issuer ?? "null"}, audience: ${this.audience ?? "null"}, verification_key: [redacted] }`;
  }
}

export class JwtClaims {
  public constructor(
    public readonly iss: string | null = null,
    public readonly sub: string | null = null,
    public readonly aud: string | null = null,
    public readonly exp: bigint | null = null,
    public readonly nbf: bigint | null = null,
    public readonly iat: bigint | null = null,
    public readonly jti: string | null = null,
    public readonly extra: Readonly<Record<string, unknown>> = {},
  ) {}

  public subject(): string | undefined {
    return this.sub ?? undefined;
  }
}

function looksLikeRsaPublicKey(value: string): boolean {
  const trimmed = value.trim();
  return (trimmed.startsWith("-----BEGIN PUBLIC KEY-----")
      && trimmed.endsWith("-----END PUBLIC KEY-----"))
    || (trimmed.startsWith("-----BEGIN RSA PUBLIC KEY-----")
      && trimmed.endsWith("-----END RSA PUBLIC KEY-----"));
}

File: sdks/python/src/echo_agent_sdk/jwt_values.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any


@dataclass(frozen=True, slots=True, repr=False)
class JwtConfig:
    _algorithm: str | None
    _verification_key: str | None = field(repr=False)
    issuer: str | None = None
    audience: str | None = None

    @classmethod
    def hs256(cls, secret: str) -> JwtConfig:
        if not isinstance(secret, str):
            raise TypeError("JWT secret must be text")
        return cls("HS256", secret)

    @classmethod
    def rs256(cls, public_key: str) -> JwtConfig:
        if not isinstance(public_key, str) or not _looks_like_rsa_public_key(
            public_key
        ):
            raise ValueError("invalid RSA public key")
        return cls("RS256", public_key)

    @classmethod
    def disabled(cls) -> JwtConfig:
        return cls(None, None)

    def with_issuer(self, issuer: str) -> JwtConfig:
        if not isinstance(issuer, str):
            raise TypeError("JWT issuer must be text")
        return JwtConfig(self._algorithm, self._verification_key, issuer, self.audience)

    def with_audience(self, audience: str) -> JwtConfig:
        if not isinstance(audience, str):
            raise TypeError("JWT audience must be text")
        return JwtConfig(self._algorithm, self._verification_key, self.issuer, audience)

    def is_enabled(self) -> bool:
        return self._algorithm is not None

    def __repr__(self) -> str:
        algorithm = self._algorithm or "none"
        return (
            "JwtConfig("
            f"enabled={self.is_enabled()}, algorithm={algorithm!r}, issuer={self.issuer!r}, "
            "audience="
            f"{self.audience!r}, verification_key='[redacted]')"
        )


@dataclass(frozen=True, slots=True)
class JwtClaims:
    iss: str | None = None
    sub: str | None = None
    aud: str | None = None
    exp: int | None = None
    nbf: int | None = None
    iat: int | None = None
    jti: str | None = None
    extra: dict[str, Any] = field(default_factory=dict)

    def subject(self) -> str | None:
        return self.sub


def _looks_like_rsa_public_key(value: str) -> bool:
    trimmed = value.strip()
    return (
        trimmed.startswith("-----BEGIN PUBLIC KEY-----")
        and trimmed.endswith("-----END PUBLIC KEY-----")
    ) or (
        trimmed.startswith("-----BEGIN RSA PUBLIC KEY-----")
        and trimmed.endswith("-----END RSA PUBLIC KEY-----")
    )
